src/daft-plan/src/optimization/rules/rule.rs

```diff
@@ -4,19 +4,59 @@ use common_error::DaftResult;
 
 use crate::LogicalPlan;
 
+/// Application order of a rule or rule batch.
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
 pub enum ApplyOrder {
+    // Apply a rule to a node and then it's children.
     TopDown,
     #[allow(dead_code)]
+    // Apply a rule to a node's children and then the node itself.
     BottomUp,
+    #[allow(dead_code)]
+    // Delegate tree traversal to the rule.
+    Delegated,
```

Comment:
From the perspective of the optimizer, kind of! The optimizer will only ever supply the root node to `ApplyOrder::Delegated` rules, but the rules are allowed to traverse and transform the entire tree within `rule.try_optimize()`. I chose the variant name `Delegated` to make it clear to someone that's reading the optimizer code that the the optimizer is handing off the tree traversal to the rule, since someone seeing `RootOnly` could think that the rule is only going to look at the root of the tree.

Definitely open to alternative perspectives on this naming though! 😄